library/ssl_tls.c

```diff
@@ -1921,6 +1921,61 @@ int mbedtls_ssl_set_hs_ecjpake_password( mbedtls_ssl_context *ssl,
 
     return( 0 );
 }
+
+int mbedtls_ssl_set_hs_ecjpake_password_opaque( mbedtls_ssl_context *ssl,
+                                                mbedtls_svc_key_id_t pwd )
+{
+    psa_pake_cipher_suite_t cipher_suite = psa_pake_cipher_suite_init();
+    psa_pake_role_t psa_role;
+    psa_status_t status;
+
+    if( ssl->handshake == NULL || ssl->conf == NULL )
+        return( MBEDTLS_ERR_SSL_BAD_INPUT_DATA );
+
+    if( ssl->conf->endpoint == MBEDTLS_SSL_IS_SERVER )
+        psa_role = PSA_PAKE_ROLE_SERVER;
+    else
+        psa_role = PSA_PAKE_ROLE_CLIENT;
+
+    if( mbedtls_svc_key_id_is_null( pwd ) )
+        return( MBEDTLS_ERR_SSL_BAD_INPUT_DATA );
+    ssl->handshake->psa_pake_password = pwd;
+
+    psa_pake_cs_set_algorithm( &cipher_suite, PSA_ALG_JPAKE );
+    psa_pake_cs_set_primitive( &cipher_suite,
+                               PSA_PAKE_PRIMITIVE( PSA_PAKE_PRIMITIVE_TYPE_ECC,
+                                                   PSA_ECC_FAMILY_SECP_R1,
+                                                   256) );
+    psa_pake_cs_set_hash( &cipher_suite, PSA_ALG_SHA_256 );
+
+    status = psa_pake_setup( &ssl->handshake->psa_pake_ctx, &cipher_suite );
+    if( status != PSA_SUCCESS )
+    {
+        psa_destroy_key( ssl->handshake->psa_pake_password );
```

Comment:
About every `psa_destroy_key` call in this function: is this something we want to do? It's not documented in the function description, and we're destroying an external key that we were not owners of.
I can understand that `mbedtls_ssl_set_hs_ecjpake_password` does that, since it's the creator of the key, but I don't think it's fine here. If someone had their own key and we're destroying it (without even describing it in the documentation), that's invasive.

Reply:
Indeed. Unless there's a good reason to deviate, for a given result (heap buffer, key store entry, etc.), the same module should be responsible for allocating and freeing.

Reply:
Sorry for this, you're completely right. Indeed yesterday afternoon I had a short discussion with @mpg who explained me that copying the key structure is different from copying the key itself (in my mind I erroneously assumed they were the same operation). Now that this is more clear to me, I confirm that this operation is a non sense.